src/intrusive_pairing_heap.rs

```diff
@@ -0,0 +1,402 @@
+    ptr::NonNull,
+};
+
+/// Compares `a` and `b` without unwinding.
+/// This is necessary to avoid reentrancy in the heap.
+fn safe_lesser<T: Ord>(a: &T, b: &T) -> bool {
+    struct DropBomb;
+    impl Drop for DropBomb {
+        fn drop(&mut self) {
+            panic!("Panicked while comparing");
+        }
+    }
+    // If `T::cmp` panics, force a double-panic (and therefore an abort).
+    let bomb = DropBomb;
+    let ordering = a < b;
+    mem::forget(bomb);
+    ordering
+}
+
+/// A node which carries data of type `T` and is stored in an intrusive heap.
+#[derive(Debug)]
+pub struct HeapNode<T> {
+    /// The parent. `None` if this is the root.
+    parent: Option<NonNull<HeapNode<T>>>,
+    /// The previous sibling. `None` if there is no previous sibling.
+    prev: Option<NonNull<HeapNode<T>>>,
+    /// The next sibling. `None` if there is no next sibling.
+    next: Option<NonNull<HeapNode<T>>>,
+    /// The first child. `None` if there are no children.
+    first_child: Option<NonNull<HeapNode<T>>>,
+    /// The data which is associated to this heap item.
+    data: T,
+    /// Prevents `HeapNode`s from being `Unpin`. They may never be moved, since
+    /// the heap semantics require addresses to be stable.
+    _pin: PhantomPinned,
+}
+
+impl<T> HeapNode<T> {
+    /// Creates a new node with the associated data
+    pub fn new(data: T) -> HeapNode<T> {
+        HeapNode::<T> {
+            parent: None,
+            prev: None,
+            next: None,
+            first_child: None,
+            data,
+            _pin: PhantomPinned,
+        }
+    }
+
+    fn is_root(&self) -> bool {
+        if self.parent.is_none() {
+            debug_assert_eq!(self.prev, None);
+            debug_assert_eq!(self.next, None);
+            true
+        } else {
+            false
+        }
+    }
+}
+
+impl<T> Deref for HeapNode<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.data
+    }
+}
+
+impl<T> DerefMut for HeapNode<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        &mut self.data
+    }
+}
+
+/// Add a child to a node.
+unsafe fn add_child<T: Ord>(
+    mut parent: NonNull<HeapNode<T>>,
+    mut child: NonNull<HeapNode<T>>,
+) {
+    // require parent <= child
+    debug_assert!(!safe_lesser(&child.as_ref().data, &parent.as_ref().data));
+    if let Some(mut old_first_child) = parent.as_mut().first_child.take() {
+        child.as_mut().next = Some(old_first_child);
+        old_first_child.as_mut().prev = Some(child);
+    }
+    parent.as_mut().first_child = Some(child);
+    child.as_mut().parent = Some(parent);
+}
+
+/// Merge two root heaps. Returns the new root.
+unsafe fn meld<T: Ord>(
+    mut left: NonNull<HeapNode<T>>,
+    mut right: NonNull<HeapNode<T>>,
+) -> NonNull<HeapNode<T>> {
+    debug_assert!(left.as_ref().is_root());
+    debug_assert!(right.as_ref().is_root());
+    if safe_lesser(&right.as_ref().data, &left.as_ref().data) {
+        mem::swap(&mut left, &mut right);
```

Comment:
This just swaps the pointers, and not their content, right?

Might be easier to understand if we just call `add_child(right, left)`  in this branch and use an `else` after it.

Reply:
True.